Improve syntax highlighting by removing dependency on c-mode

Wilfred opened the pull request:
This changes syntax highlighting in hack-mode to be a standalone mode without using c-mode. This simplifies the logic, as we don't need to tell Emacs to ignore c-mode features that aren't applicable to hack-mode, such as the C preprocessor.

It also fixes Emacs 26 support, as some c-mode APIs have changed (e.g. c-font-lock-declarators now takes another argument) and would reliably crash when opening Hack files:

    Debugger entered--Lisp error: (wrong-type-argument stringp nil)
      search-forward-regexp(nil 138 bound)
      c-depropertize-CPP(1 138)
      #f(compiled-function (fn) #<bytecode 0x41f57035>)(c-depropertize-CPP)
      mapc(#f(compiled-function (fn) #<bytecode 0x41f57035>) (c-extend-region-for-CPP c-depropertize-CPP c-invalidate-macro-cache c-truncate-bs-cache c-parse-quotes-before-change))
      c-common-init(hack-mode)
      (let ((delay-mode-hooks t)) (c-mode) (setq major-mode 'hack-mode) (setq mode-name "Hack") (progn (if (get 'c-mode 'mode-class) (put 'hack-mode 'mode-class (get 'c-mode 'mode-class))) (if (keymap-parent hack-mode-map) nil (set-keymap-parent hack-mode-map (current-local-map))) (let ((parent (char-table-parent hack-mode-syntax-table))) (if (and parent (not (eq parent (standard-syntax-table)))) nil (set-char-table-parent hack-mode-syntax-table (syntax-table)))) (if (or (abbrev-table-get hack-mode-abbrev-table :parents) (eq hack-mode-abbrev-table local-abbrev-table)) nil (abbrev-table-put hack-mode-abbrev-table :parents (list local-abbrev-table)))) (use-local-map hack-mode-map) (set-syntax-table hack-mode-syntax-table) (setq local-abbrev-table hack-mode-abbrev-table) (c-initialize-cc-mode t) (funcall (function (lambda nil (require 'cc-langs) (let ((c-buffer-is-cc-mode 'hack-mode) (init (append (cdr c-emacs-variable-inits) (cdr c-lang-variable-inits))) current-var) (progn (make-local-variable 'comment-start) (make-local-variable 'comment-end) (make-local-variable 'comment-start-skip) (make-local-variable 'comment-end-can-be-escaped) (make-local-variable 'beginning-of-defun-function) (make-local-variable 'end-of-defun-function)) (condition-case err (let ((--dolist-tail-- init)) (while --dolist-tail-- (let ((var-init (car --dolist-tail--))) (setq current-var (car var-init)) (set (car var-init) (eval (car (cdr var-init)))) (setq --dolist-tail-- (cdr --dolist-tail--))))) (error (if current-var (message "Eval error in the `c-lang-defvar' or `c-lang-setver' for `%s' (source eval): %S" current-var err) (signal (car err) (cdr err))))))))) (c-common-init 'hack-mode) (set (make-local-variable 'font-lock-maximum-decoration) t) (set (make-local-variable 'case-fold-search) t) (set (make-local-variable 'compile-command) (concat hack-client-program-name " --from emacs")) (set (make-local-variable 'indent-line-function) (function hack-xhp-indent-line)) (c-set-style "hack") (run-mode-hooks 'c-mode-common-hook 'hack-mode-hook) (c-update-modeline))
      (progn (make-local-variable 'delay-mode-hooks) (let ((delay-mode-hooks t)) (c-mode) (setq major-mode 'hack-mode) (setq mode-name "Hack") (progn (if (get 'c-mode 'mode-class) (put 'hack-mode 'mode-class (get 'c-mode 'mode-class))) (if (keymap-parent hack-mode-map) nil (set-keymap-parent hack-mode-map (current-local-map))) (let ((parent (char-table-parent hack-mode-syntax-table))) (if (and parent (not (eq parent (standard-syntax-table)))) nil (set-char-table-parent hack-mode-syntax-table (syntax-table)))) (if (or (abbrev-table-get hack-mode-abbrev-table :parents) (eq hack-mode-abbrev-table local-abbrev-table)) nil (abbrev-table-put hack-mode-abbrev-table :parents (list local-abbrev-table)))) (use-local-map hack-mode-map) (set-syntax-table hack-mode-syntax-table) (setq local-abbrev-table hack-mode-abbrev-table) (c-initialize-cc-mode t) (funcall (function (lambda nil (require 'cc-langs) (let ((c-buffer-is-cc-mode 'hack-mode) (init (append (cdr c-emacs-variable-inits) (cdr c-lang-variable-inits))) current-var) (progn (make-local-variable 'comment-start) (make-local-variable 'comment-end) (make-local-variable 'comment-start-skip) (make-local-variable 'comment-end-can-be-escaped) (make-local-variable 'beginning-of-defun-function) (make-local-variable 'end-of-defun-function)) (condition-case err (let ((--dolist-tail-- init)) (while --dolist-tail-- (let ((var-init (car --dolist-tail--))) (setq current-var (car var-init)) (set (car var-init) (eval (car (cdr var-init)))) (setq --dolist-tail-- (cdr --dolist-tail--))))) (error (if current-var (message "Eval error in the `c-lang-defvar' or `c-lang-setver' for `%s' (source eval): %S" current-var err) (signal (car err) (cdr err))))))))) (c-common-init 'hack-mode) (set (make-local-variable 'font-lock-maximum-decoration) t) (set (make-local-variable 'case-fold-search) t) (set (make-local-variable 'compile-command) (concat hack-client-program-name " --from emacs")) (set (make-local-variable 'indent-line-function) (function hack-xhp-indent-line)) (c-set-style "hack") (run-mode-hooks 'c-mode-common-hook 'hack-mode-hook) (c-update-modeline)))
      hack-mode()
      funcall-interactively(hack-mode)
      call-interactively(hack-mode record nil)
      command-execute(hack-mode record)

Syntax highlighting is now based on the lexer used by Hack, ensuring that we have full coverage (e.g. nonnull was missing before). Variables $foo are highlighted, but the $ itself is not, to make it easy to focus on the name (the same as php-mode).

Function and method names are now also highlighted.

Wilfred commented:
Before and after on Emacs 26:

![screen shot 2018-09-07 at 10 49 50](https://user-images.githubusercontent.com/70800/45213245-e44aaf00-b28e-11e8-8ff8-77642b5e00e0.png)

![screen shot 2018-09-07 at 10 50 17](https://user-images.githubusercontent.com/70800/45213249-e90f6300-b28e-11e8-8497-840a8ba5c598.png)

jra3 commented on hack-mode.el:
```diff
@@ -43,589 +38,225 @@
-(c-add-style "hack" hack-style)
+(defvar hack-mode-syntax-table
+  (let ((table (make-syntax-table)))
+    ;; - and < are not symbol constituents.
```
I wonder if you might know how to work around this... In XHP, '-' does appear in names often. But this is only XHP class and attribute names, not the rest of the language. It's not a showstopper, but it would be cool to have a good solution for XHP (especially since I write a lot of it :) )

Wilfred replied:
I think we can do something smart here. If we put a syntax code on occurrences of `<` that signify the beginning of an XHP expression, we can then mark any `-` as a symbol constituent up until the closing `>`.

I haven't had a chance to look at the XHP parser yet, but it is something I want to do :)

jra3 commented on hack-mode.el:
```diff
@@ -43,589 +38,225 @@
 
-  (setq-local font-lock-maximum-decoration t)
-  (setq-local case-fold-search t)
-  (setq-local compile-command (concat hack-client-program-name " --from emacs"))
```
why remove that?

Wilfred replied:
Woops! Carelessness.

jra3 commented:
This seems to remove the 3 specialized indentation functions fixing arglists and continuations of function chains inside of arrays/dicts. I'll install this version and check out how different the indenting feels.

jra3 commented:
font lockinging for the $$ symbol is funky. I think making them the same face as something like... the ==> operator is good.

jra3 commented:
We can do other font-lock improvements as followups, but indentation should probably be fixed.

Wilfred commented:
@jra3 OK, I've updated the PR:

* Highlight of `$$` now highlights the whole variable, to make it visibly distinct form `$s` but still use the same colours.
* Fixed a few syntax issues with `>` and `+` that I spotted when testing this.
* Fixed up indentation to be much less reliant on cc-mode.

Let me know what you think :). You may need to try this code in a new Emacs instance: it's a pain to `makunbound` the mode-map, and if you don't then you end up with keybindings from c-mode sticking around.

jra3 commented:
Much simpler :)

There's some work to do with indentation, but this highlighting is much, much better.
